### PR TITLE
chore(logstreams): add checksums to the snapshot chunks

### LINKED ISSUE #2345: Verify that a snapshot is replicated successfully
AT:
* when a snapshot is replicated to a follower then the follower verifies that the snapshot is replicated  successfully (e.g. checksum, open snapshot, etc.)
* if the snapshot is correct then it is marked as valid
* otherwise, it is discarded

requires #2344

### LINKED ISSUE #2378: Snapshot replication could in some situation mistakenly assume to early that snapshot is valid
Currently we sends each snapshot file as a separate chunk. After receiving this chunk we write the content to a new file. After writing this file we check how many files exist in the current snapshot temporary folder, if the count is equal to the total count (total count is transmitted with the snapshot chunk), then we assume this snapshot is valid.

It could happen that there already exist files in this directory or someone added files during the replication, then the assumption that we received all files is wrong and we should not move the snapshot to a valid state.

### PR DESCRIPTION
**Pull request by ChrisKujawa**
* checksums are created before the chunk is replicated
 * checksum is validated on receiving chunk
 * if sum is not equal we will mark this snapshot as invalid
 * we count the received chunks with a in mem map
 * we ensure max snapshot count after marking snapshot as valid 

closes #2345
closes #2378

follow up issue https://github.com/zeebe-io/zeebe/issues/2387

### CODE REVIEW

File: logstreams/src/main/java/io/zeebe/logstreams/state/ReplicationController.java [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+
+    LOG.debug("Consume snapshot chunk {}", chunkName);
+    writeReceivedSnapshotChunk(snapshotChunk, tmpSnapshotDirectory, snapshotFile);
+  }
+
+  private long createActualChecksum(SnapshotChunk snapshotChunk) {
+    final byte[] content = snapshotChunk.getContent();
+    final CRC32 crc32 = new CRC32();
+    crc32.update(content);
+    return crc32.getValue();
```
Comment (saig0) [inline logstreams/src/main/java/io/zeebe/logstreams/state/ReplicationController.java]: I would prefer to use the same method to create the checksums at the sender and the receiver.

File: logstreams/src/main/java/io/zeebe/logstreams/state/StateSnapshotController.java [hunk 2]
```diff
@@ -38,19 +34,32 @@
...
+      StateStorage storage,
+      SnapshotReplication replication,
+      int maxSnapshots) {
     this.storage = storage;
-    this.replication = replication;
     this.zeebeDbFactory = zeebeDbFactory;
+    replicationController =
+        new ReplicationController(
+            replication,
+            storage,
```
Comment (saig0) [inline logstreams/src/main/java/io/zeebe/logstreams/state/StateSnapshotController.java]: I get confused about all the different names: `StateSnapshotControler`, `SnapshotReplication`, `ReplicationController`, `AsyncSnapshotDirector`, etc. :sweat_smile:

File: logstreams/src/test/java/io/zeebe/logstreams/state/NotReplicateAllChunksTest.java [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    receiverSnapshotController =
+        new StateSnapshotController(
+            ZeebeRocksDbFactory.newFactory(DefaultColumnFamily.class),
+            receiverStorage,
+            replicator,
+            1);
+
+    autoCloseableRule.manage(replicatorSnapshotController);
+    autoCloseableRule.manage(receiverSnapshotController);
+    replicatorSnapshotController.openDb();
```
Comment (saig0) [inline logstreams/src/test/java/io/zeebe/logstreams/state/NotReplicateAllChunksTest.java]: The setup is similar to `InvalidChecksumSnapshotChunkTest`. Maybe, we can merge both tests together. Just move the test method and the replicator.

**Review by saig0 — CHANGES_REQUESTED** [review]
Very good :+1: Just a few notes.

**Comment by Zelldon** [comment]
@saig0 ping

**Review by saig0 — APPROVED** [review]

**Comment by Zelldon** [comment]
Bors r+

**Pull request merged**